File: src/models/command.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// HTTP method enum
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum HttpMethod {
    GET,
    POST,
    PUT,
    DELETE,
    PATCH,
    HEAD,
    OPTIONS,
    TRACE,
    CONNECT,
}

impl Default for HttpMethod {
    fn default() -> Self {
        Self::GET
    }
}

impl std::fmt::Display for HttpMethod {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            HttpMethod::GET => write!(f, "GET"),
            HttpMethod::POST => write!(f, "POST"),
            HttpMethod::PUT => write!(f, "PUT"),
            HttpMethod::DELETE => write!(f, "DELETE"),
            HttpMethod::PATCH => write!(f, "PATCH"),
            HttpMethod::HEAD => write!(f, "HEAD"),
            HttpMethod::OPTIONS => write!(f, "OPTIONS"),
            HttpMethod::TRACE => write!(f, "TRACE"),
            HttpMethod::CONNECT => write!(f, "CONNECT"),
        }
    }
}

/// Header struct
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Header {
    /// Unique identifier
    pub id: String,
    /// Header key
    pub key: String,
    /// Header value
    pub value: String,
    /// Whether the header is enabled
    pub enabled: bool,
}

/// Query parameter struct
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryParam {
    /// Unique identifier
    pub id: String,
    /// Parameter key
    pub key: String,
    /// Parameter value
    pub value: String,
    /// Whether the parameter is enabled
    pub enabled: bool,
}

/// Form data item struct
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FormDataItem {
    /// Unique identifier
    pub id: String,
    /// Item key
    pub key: String,
    /// Item value
    pub value: String,
    /// Whether the item is enabled
    pub enabled: bool,
}

/// Request body enum
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum RequestBody {
    /// No body
    None,
    /// Raw text body
    Raw(String),
    /// Form data body
    FormData(Vec<FormDataItem>),
    /// Binary file body
    Binary(PathBuf),
}

impl Default for RequestBody {
    fn default() -> Self {
        Self::None
    }
}

/// Curl option struct
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CurlOption {
    /// Unique identifier
    pub id: String,
    /// Option flag (e.g., "-v", "--location")
    pub flag: String,
    /// Option value (if applicable)
    pub value: Option<String>,
    /// Whether the option is enabled
    pub enabled: bool,
}

/// Curl command struct
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CurlCommand {
    /// Unique identifier
    pub id: String,
    /// Command name
    pub name: String,
    /// Command description
    pub description: Option<String>,
    /// URL
    pub url: String,
    /// HTTP method
    pub method: Option<HttpMethod>,
    /// Headers
    pub headers: Vec<Header>,
    /// Query parameters
    pub query_params: Vec<QueryParam>,
    /// Request body
    pub body: Option<RequestBody>,
    /// Curl options
    pub options: Vec<CurlOption>,
    /// Creation timestamp
    pub created_at: DateTime<Utc>,
    /// Last update timestamp
    pub updated_at: DateTime<Utc>,
}

impl Default for CurlCommand {
    fn default() -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name: "New Command".to_string(),
            description: None,
            url: "https://".to_string(),
            method: Some(HttpMethod::GET),
            headers: Vec::new(),
            query_params: Vec::new(),
            body: None,
            options: Vec::new(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }
}

impl CurlCommand {
    /// Create a new curl command
    pub fn new(url: String) -> Self {
        Self {
            url,
            ..Self::default()
        }
    }

    /// Add a header
    pub fn add_header(&mut self, key: String, value: String) -> &mut Self {
        self.headers.push(Header {
            id: uuid::Uuid::new_v4().to_string(),
            key,
            value,
            enabled: true,
        });
        self
    }

    /// Add a query parameter
    pub fn add_query_param(&mut self, key: String, value: String) -> &mut Self {
        self.query_params.push(QueryParam {
            id: uuid::Uuid::new_v4().to_string(),
            key,
            value,
            enabled: true,
        });
        self
    }

    /// Add a curl option
    pub fn add_option(&mut self, flag: String, value: Option<String>) -> &mut Self {
        self.options.push(CurlOption {
            id: uuid::Uuid::new_v4().to_string(),
            flag,
            value,
            enabled: true,
        });
        self
    }

    /// Set the HTTP method
    pub fn set_method(&mut self, method: HttpMethod) -> &mut Self {
        self.method = Some(method);
        self
    }

    /// Set the request body
    pub fn set_body(&mut self, body: RequestBody) -> &mut Self {
        self.body = Some(body);
        self
    }
}

File: src/zvrl/core/models/command.zig
const std = @import("std");
const ids = @import("../ids.zig");

const Allocator = std.mem.Allocator;

pub const HttpMethod = enum {
    get,
    post,
    put,
    delete,
    patch,
    head,
    options,
    trace,
    connect,

    pub fn default() HttpMethod {
        return .get;
    }

    pub fn asString(self: HttpMethod) []const u8 {
        return switch (self) {
            .get => "GET",
            .post => "POST",
            .put => "PUT",
            .delete => "DELETE",
            .patch => "PATCH",
            .head => "HEAD",
            .options => "OPTIONS",
            .trace => "TRACE",
            .connect => "CONNECT",
        };
    }
};

pub const Header = struct {
    id: u64,
    key: []u8,
    value: []u8,
    enabled: bool,

    pub fn deinit(self: *Header, allocator: Allocator) void {
        allocator.free(self.key);
        allocator.free(self.value);
    }
};

pub const QueryParam = struct {
    id: u64,
    key: []u8,
    value: []u8,
    enabled: bool,

    pub fn deinit(self: *QueryParam, allocator: Allocator) void {
        allocator.free(self.key);
        allocator.free(self.value);
    }
};

pub const FormDataItem = struct {
    id: u64,
    key: []u8,
    value: []u8,
    enabled: bool,

    pub fn deinit(self: *FormDataItem, allocator: Allocator) void {
        allocator.free(self.key);
        allocator.free(self.value);
    }
};

pub const RequestBody = union(enum) {
    none,
    raw: []u8,
    form_data: std.ArrayList(FormDataItem),
    binary: []u8,

    pub fn deinit(self: *RequestBody, allocator: Allocator) void {
        switch (self.*) {
            .none => {},
            .raw => |value| allocator.free(value),
            .binary => |value| allocator.free(value),
            .form_data => |*list| {
                for (list.items) |*item| {
                    item.deinit(allocator);
                }
                list.deinit(allocator);
            },
        }
    }
};

pub const CurlOption = struct {
    id: u64,
    flag: []u8,
    value: ?[]u8,
    enabled: bool,

    pub fn deinit(self: *CurlOption, allocator: Allocator) void {
        allocator.free(self.flag);
        if (self.value) |value| allocator.free(value);
    }
};

pub const CurlCommand = struct {
    allocator: Allocator,
    id: u64,
    name: []u8,
    description: ?[]u8,
    url: []u8,
    method: ?HttpMethod,
    headers: std.ArrayList(Header),
    query_params: std.ArrayList(QueryParam),
    body: ?RequestBody,
    options: std.ArrayList(CurlOption),
    created_at: ids.Timestamp,
    updated_at: ids.Timestamp,

    pub fn init(allocator: Allocator, generator: *ids.IdGenerator) !CurlCommand {
        const now = ids.nowTimestamp();
        return .{
            .allocator = allocator,
            .id = generator.nextId(),
            .name = try allocator.dupe(u8, "New Command"),
            .description = null,
            .url = try allocator.dupe(u8, "https://"),
            .method = HttpMethod.default(),
            .headers = try std.ArrayList(Header).initCapacity(allocator, 0),
            .query_params = try std.ArrayList(QueryParam).initCapacity(allocator, 0),
            .body = null,
            .options = try std.ArrayList(CurlOption).initCapacity(allocator, 0),
            .created_at = now,
            .updated_at = now,
        };
    }

    pub fn new(allocator: Allocator, generator: *ids.IdGenerator, url: []const u8) !CurlCommand {
        var command = try CurlCommand.init(allocator, generator);
        allocator.free(command.url);
        command.url = try allocator.dupe(u8, url);
        return command;
    }

    pub fn deinit(self: *CurlCommand) void {
        self.allocator.free(self.name);
        if (self.description) |value| self.allocator.free(value);
        self.allocator.free(self.url);

        if (self.body) |*body| {
            body.deinit(self.allocator);
        }

        for (self.headers.items) |*header| {
            header.deinit(self.allocator);
        }
        self.headers.deinit(self.allocator);

        for (self.query_params.items) |*param| {
            param.deinit(self.allocator);
        }
        self.query_params.deinit(self.allocator);

        for (self.options.items) |*option| {
            option.deinit(self.allocator);
        }
        self.options.deinit(self.allocator);
    }

    pub fn addHeader(self: *CurlCommand, generator: *ids.IdGenerator, key: []const u8, value: []const u8) !void {
        try self.headers.append(self.allocator, .{
            .id = generator.nextId(),
            .key = try self.allocator.dupe(u8, key),
            .value = try self.allocator.dupe(u8, value),
            .enabled = true,
        });
        self.updated_at = ids.nowTimestamp();
    }

    pub fn addQueryParam(self: *CurlCommand, generator: *ids.IdGenerator, key: []const u8, value: []const u8) !void {
        try self.query_params.append(self.allocator, .{
            .id = generator.nextId(),
            .key = try self.allocator.dupe(u8, key),
            .value = try self.allocator.dupe(u8, value),
            .enabled = true,
        });
        self.updated_at = ids.nowTimestamp();
    }

    pub fn addOption(self: *CurlCommand, generator: *ids.IdGenerator, flag: []const u8, value: ?[]const u8) !void {
        const owned_value = if (value) |item| try self.allocator.dupe(u8, item) else null;
        try self.options.append(self.allocator, .{
            .id = generator.nextId(),
            .flag = try self.allocator.dupe(u8, flag),
            .value = owned_value,
            .enabled = true,
        });
        self.updated_at = ids.nowTimestamp();
    }

    pub fn setMethod(self: *CurlCommand, method: HttpMethod) void {
        self.method = method;
        self.updated_at = ids.nowTimestamp();
    }

    pub fn setBody(self: *CurlCommand, body: RequestBody) void {
        if (self.body) |*existing| {
            existing.deinit(self.allocator);
        }
        self.body = body;
        self.updated_at = ids.nowTimestamp();
    }
};

test "curl command default values" {
    var generator = ids.IdGenerator{};
    var command = try CurlCommand.init(std.testing.allocator, &generator);
    defer command.deinit();

    try std.testing.expectEqualStrings("New Command", command.name);
    try std.testing.expectEqualStrings("https://", command.url);
    try std.testing.expect(command.method.? == .get);
    try std.testing.expect(command.headers.items.len == 0);
    try std.testing.expect(command.options.items.len == 0);
}

test "curl command add header/option" {
    var generator = ids.IdGenerator{};
    var command = try CurlCommand.init(std.testing.allocator, &generator);
    defer command.deinit();

    try command.addHeader(&generator, "Accept", "application/json");
    try command.addOption(&generator, "-v", null);

    try std.testing.expectEqual(@as(usize, 1), command.headers.items.len);
    try std.testing.expectEqual(@as(usize, 1), command.options.items.len);
}
